File: solutions/typescript/largest-series-product/1/largest-series-product.ts
export const largestProduct = (digits: string, span: number): number => {
  if (span < 0) throw new Error('Span must not be negative');
  if (span > digits.length) throw new Error('Span must not exceed string length');
  if (/\D/.test(digits)) throw new Error('Digits input must only contain digits');
  
  if (span === 0) return 1;

  let maxProduct = 0;

  for (let i = 0; i <= digits.length - span; i++) {
    const series = digits.substring(i, i + span);
    const product = series
      .split('')
      .reduce((acc, digit) => acc * parseInt(digit, 10), 1);

    if (product > maxProduct) {
      maxProduct = product;
    }
  }

  return maxProduct;
};

File: solutions/kotlin/largest-series-product/1/src/main/kotlin/LargestSeriesProduct.kt
class Series(private val input: String) {

    init {
        require(input.all { it.isDigit() }) { "Input must only contain digits." }
    }

    fun getLargestProduct(span: Int): Long {
        require(span in 0..input.length) { "Span must be smaller than or equal to string length and non-negative." }
        if (span == 0) return 1L

        return input.windowed(span) { window ->
            window.map { it.toString().toLong() }.reduce { acc, digit -> acc * digit }
        }.maxOrNull() ?: 1L
    }
}
